### ISSUE TITLE
Proposal: adding a separate PointDataBuilder for PointData

### ISSUE THREAD

**Issue by ldematte** [influxdata/influxdb-client-csharp#178]
I see that PointData is coded to be, more or less, immutable. Any operations like Tag or Field returns a new PointData object. As a general principle, I like immutable objects for the guarantees they offer, but may I ask you why is it used in this case?
Anyway, the downside is that any modification results in a new object, so if (for example) I have

`var point = Point.Measurement(..).Tag(...).Tag(...).Field(....).Timestamp(...)`

I have generated 5 objects while I end up using just 1. This puts a lot of pressure on the GC, and this code is potentially executed at a high rate, so why not use a separate builder? Very much like StringBuilder for strings.
What do you guys think?

**Comment by bednar** [comment]
> I see that PointData is coded to be, more or less, immutable. Any operations like Tag or Field returns a new PointData object. As a general principle, I like immutable objects for the guarantees they offer, but may I ask you why is it used in this case?

The immutable `PointData` offers possibility to reuse point in different writes. Something like:

```csharp
var pointBase = PointData.Measurement("xyz").Tag("company", "best-one").Tag("feature", "a");
```

```csharp
var point = pointBase.Field("count", 10); 
using (var writeApi = _client.GetWriteApi()) { 
   writeApi.WritePoint(BUCKET, TENANT, point); 
}
```

> Anyway, the downside is that any modification results in a new object, so if (for example) I have
> 
> `var point = Point.Measurement(..).Tag(...).Tag(...).Field(....).Timestamp(...)`
> 
> I have generated 5 objects while I end up using just 1. This puts a lot of pressure on the GC, and this code is potentially executed at a high rate, so why not use a separate builder? Very much like StringBuilder for strings.

How will be api looks like? What about something like `PointDataMutable`?

> What do you guys think?

It is correct requirement... let's do it. Is this something you might be willing to help with?

**Comment by ldematte** [comment]
> How will be api looks like? What about something like `PointDataMutable`?

I would keep `PointData` immutable and add a `PointDataBuilder`, very much like `string`/`StringBuilder` and the various [Immutable collection classes](https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable?view=net-5.0) 

For example:
```cs
var builder = PointData.Builder.Measurement("xyz").Tag("company", "best-one").Tag("feature", "a");
PointData pointData = builder.ToPoint();
```
Internally, `PointDataBuilder` will have mutable (private) fields (to avoid allocations) and then use the private `PointData` ctor when ToPoint is called to produce a new immutable PointData.

The new api would be parallel to the existing one, so you can choose which approach is convenient for your scenario (again, very much like `string`/`StringBuilder` or `ImmutableList`/`ImmutableList.Builder`)

> 
> It is correct requirement... let's do it. Is this something you might be willing to help with?

Sure, would love to help!
Let me do the CancellationToken one first, I probably have time to do it tomorrow, and then I can do this.

**Comment by bednar** [comment]
> > How will be api looks like? What about something like `PointDataMutable`?
> 
> I would keep `PointData` immutable and add a `PointDataBuilder`, very much like `string`/`StringBuilder` and the various [Immutable collection classes](https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable?view=net-5.0)
> 
> For example:
> 
> ```cs
> var builder = PointData.Builder.Measurement("xyz").Tag("company", "best-one").Tag("feature", "a");
> PointData pointData = builder.ToPoint();
> ```
> 

Nice API 👍 

> Internally, `PointDataBuilder` will have mutable (private) fields (to avoid allocations) and then use the private `PointData` ctor when ToPoint is called to produce a new immutable PointData.
> 
> The new api would be parallel to the existing one, so you can choose which approach is convenient for your scenario (again, very much like `string`/`StringBuilder` or `ImmutableList`/`ImmutableList.Builder`)
> 
> > It is correct requirement... let's do it. Is this something you might be willing to help with?
> 
> Sure, would love to help!
> Let me do the CancellationToken one first, I probably have time to do it tomorrow, and then I can do this.

Thanks a lot of for your help.

**Comment by bednar** [comment]
The issue is fixed in [1.17.0](https://github.com/influxdata/influxdb-client-csharp/milestone/20) milestone. You could use the preview version by:

```
dotnet add package InfluxDB.Client --version 1.17.0-dev.2652
```

https://www.nuget.org/packages/InfluxDB.Client/1.17.0-dev.2652
